rust/src/index/vector/diskann/builder.rs

```diff
@@ -0,0 +1,494 @@
+            }
+        }
+
+        // Make bidirectional connections.
+        {
+            let n = graph.neighbors_mut(i);
+            n.clear();
+            n.extend(neighbor_ids.iter().copied());
+        }
+        {
+            for neighbor_id in neighbor_ids.iter() {
+                graph.add_edge(*neighbor_id as usize, i);
+            }
+        }
+    }
+
+    Ok((graph, matrix))
+}
+
+/// Distance from the query vector to the vector at the given idx.
+fn distance_to(vectors: &MatrixView, query: &[f32], idx: usize) -> Result<f32> {
+    let vector = vectors
+        .row(idx)
+        .ok_or(Error::Index("Invalid row index".to_string()))?;
+
+    let dists = l2_distance(query, vector, vectors.num_columns())?;
+    Ok(dists.value(0))
+}
+
+/// Distance between two vectors in the matrix.
+fn distance(matrix: &MatrixView, i: usize, j: usize) -> Result<f32> {
+    let vector_i = matrix
+        .row(i)
+        .ok_or(Error::Index("Invalid row index".to_string()))?;
+    let vector_j = matrix
+        .row(j)
+        .ok_or(Error::Index("Invalid row index".to_string()))?;
+
+    let dists = l2_distance(vector_i, vector_j, matrix.num_columns())?;
+    Ok(dists.value(0))
+}
+
+/// Greedy search.
+///
+/// Algorithm 1 in the paper.
+///
+/// Parameters:
+/// - start: The starting vertex.
+/// - query: The query vector.
+/// - k: The number of nearest neighbors to return.
+/// - search_size: Search list size, L in the paper.
+async fn greedy_search(
+    graph: &GraphBuilder<PQVertex, DiskANNParams>,
+    vectors: &MatrixView,
+    start: usize,
+    query: &[f32],
+    k: usize,
+    search_size: usize, // L in the paper.
+) -> Result<(Vec<usize>, HashSet<usize>)> {
+    let mut visited: HashSet<usize> = HashSet::new();
+
+    // L in the paper.
+    // A map from distance to vertex id.
+    let mut candidates: BTreeMap<OrderedFloat<f32>, usize> = BTreeMap::new();
+    let mut heap: BinaryHeap<VertexWithDistance> = BinaryHeap::new();
+    let dist = distance_to(vectors, query, start)?;
+    heap.push(VertexWithDistance {
+        id: start,
+        distance: OrderedFloat(dist),
+    });
+    candidates.insert(OrderedFloat(distance_to(vectors, query, start)?), start);
+    while let Some(p) = heap.pop() {
+        // In paper:
+        // p = argmin_{L \ V} d(p, q)
+        if visited.contains(&p.id) || !candidates.contains_key(&p.distance) {
+            continue;
+        }
+        visited.insert(p.id);
+        for neighbor_id in graph.neighbors(p.id).iter() {
+            let neighbor_id = *neighbor_id as usize;
+            if visited.contains(&neighbor_id) {
+                // Already visited.
+                continue;
+            }
+            let dist = distance_to(vectors, query, neighbor_id)?;
+            candidates.insert(OrderedFloat(dist), neighbor_id as usize);
+            if candidates.len() > search_size {
+                candidates.pop_last();
+            }
+        }
+    }
+
+    Ok((
```

Comment:
so currently this is searching through all neighbors of the starting vertex and returning the K closest?

Reply:
yes.